processor/processor.go

```diff
@@ -19,16 +19,27 @@ import (
 	"github.com/open-telemetry/opentelemetry-collector/consumer"
 )
 
+// Processor defines the common functions that must be implemented by TraceProcessor
+// and MetricsProcessor.
+type Processor interface {
+	// GetCapabilities must return the capabilities of the processor.
+	GetCapabilities() Capabilities
+
+	// Shutdown is invoked during service shutdown. Typically used to flush the data
+	// that may be accumulated in the processor.
+	Shutdown() error
```

Comment:
Do we need shutdown on all the Consumer maybe? In that case you call shutdown on the receiver and then the receiver calls the next shutdown? May have problems in case of fanin processors.

Reply:
>May have problems in case of fanin processors.

Yes, chained Shutdown calls will create problems with fanout connectors. It will mean we may call Shutdown() more than once (which creates additional burden for all Shutdown() implementations).

I want to have explicit Shutdown() for processors and implement proper shutdown sequence in `Application.shutdownPipelines()` (concurrently when there are multiple pipelines).

Reply:
Exporters already have a Shutdown method, after adding it to processors all consumers will have this method. Why not have the Shutdown at the consumer interfaces instead? Churn in tests because of cascading changes?

Reply:
Not all consumers need a Shutdown method, processors and exporters are only a subset of all consumers in our codebase. It will result in unnecessary empty Shutdown() implementations in the code (for example 4 have Fanout connectors which are also consumers).